### PR TITLE
197 pytree regularizer strength

### PR DESCRIPTION
**Pull request by wulfdewolf**
# Finer control over regularization strength

## Summary 
This PR extends NeMoS’ regularization API to allow per-parameter control of regularization strength. In addition to scalar strengths, users can now supply pytrees matching the model parameter structure to regularize different parameters (or groups) with different strengths. `ElasticNet` now supports a tuple (strength, ratio) where either or both can be pytrees. `GroupLasso` now allows per-group strengths.

## Motivation
- When fitting a `PopulationGLM`, individual neurons can be regularized independently.
- When new models are added, we will want more fine control over which parameters are regularized.

## Key changes

- Accept strength as pytree
  * `regularizer_strength` now accepts: `None` (defaults appropriately per-regularizer), scalar (float / 0-D array), or a pytree matching the structure of the regularizable parameters.
  * For ElasticNet: a tuple where each of `(strength, ratio)` can be a scalar or pytree. 
  * Regressor signatures updated accordingly (typing changed to `Any`). 

- Regularizer interface refactor
  * `proximal_operator` field, allowing for a superclass `get_proximal_operator`
  * `get_proximal_operator(params, strength)`: now closes over strength (still possible to pass hyperparameters to proximal operator, but ignored for now).
  * `penalized_loss(loss, params, strength)`: embeds penalty with structured strengths.
  * New validation helpers:
     - `_validate_strength`: converts scalars or trees to JAX arrays; defaults `None` as appropriate.
     - `_validate_strength_structure`: broadcasts/aligns strengths to regularizable parameter subtrees and inserts `None` elsewhere.
  * `_get_filter_kwargs(params, strength)`: standardized way to pass masked/broadcast strengths into penalties and proximal operators.
  * Penalty computation now tree-maps over `(params, strength)` pairs; intercepts remain unregularized via `regularizable_subtrees()`, as before.

- Proximal operators generalized, all `prox_` functions now accept a `strength` pytree
  * `prox_none(x, strength, scaling)`
  * `prox_ridge(x, strength, scaling)`: divides by `(1 + scaling * strength)` leafwise.
  * `prox_lasso(x, strength, scaling)`: soft-threshold leafwise by `scaling * strength`.
  * `prox_elastic_net(x, strength, scaling)`: expects per-leaf `(strength, ratio)`; applies combined L1/L2 shrinkage.
  * `prox_group_lasso(x, strength, mask, scaling)`: expects per-group strengths (vector); applies group-wise shrinkage.

- Solver integration
  * `JaxoptAdapter` and `OptimistixAdapter` updated to embed the strength into the proximal operator via `get_proximal_operator(params, strength)`, similarly `penalized_loss(loss, params, strength)` for non-proximal methods. 
  * `FISTA`/`GradientDescent` no longer have a `regularizer_strength` field. They now call `prox(new_y, (), stepsize)` since strength is now embedded and we have no other hyperparameters `()`.

- Utilities
  * in `type_casting.py`: `_is_scalar_or_0d` helper.
  * in `tree_utils.py`: `tree_full_like` to quickly build pytree strengths for tests/usage.
  * `io.inspect_npz`: nicer float formatting for metadata and model attributes.

## User-facing API changes

You can still pass a scalar as before; existing scripts continue to work.
You can now also pass:
- A pytree of strengths matching model parameters (e.g., GLMParams-like trees).
- For `ElasticNet`, a tuple `(strength, ratio)` where each element can be scalar or pytree.
- For `GroupLasso`, a scalar or a 1D array matching the number of groups.

## Examples
- Scalar strength (unchanged):
```
X = nmo.pytrees.FeaturePytree(
    f1=np.random.normal(size=(100, 5)),
    f2=np.random.normal(size=(100, 2)),
)
y = np.random.randint(low=0, high=10, size=(100,))
glm = nmo.glm.GLM(regularizer="Ridge", regularizer_strength=0.1)
glm.fit(X, y)
```
- Parameter group-wise strength:
```
glm = nmo.glm.GLM(
    regularizer="Lasso", 
    regularizer_strength=dict(f1=1.0, f2=1.0),
)
glm.fit(X, y)
```
- Parameter-wise strength:
```
glm = nmo.glm.GLM(
    regularizer="Lasso", 
    regularizer_strength=dict(f1=[1.0, 1.0, 1.0, 0.5, 0.5], f2=[1.0, 0.4]),
)
glm.fit(X, y)
```
- `ElasticNet` mixing parameter group-wise and parameter-wise strengths and ratios:
```
glm = nmo.glm.GLM(
    regularizer="ElasticNet", 
    regularizer_strength=(
        dict( # strength
            f1=[1.0, 1.0, 1.0, 0.5, 0.5], 
            f2=[1.0, 0.4]
        ),
        dict( # ratio
            f1=0.3, 
            f2=[0.4, 0.6]
        ),
    )
)
glm.fit(X, y)
```
- `GroupLasso` with per-group strengths:
```
glm = nmo.glm.GLM(
    regularizer="GroupLasso", 
    regularizer_strength=[0.2, 1.0],
)
glm.fit(X, y)
```

- `PopulationGLM` with per-neuron (or even per-parameter) strengths:
```
y = np.random.randint(low=0, high=10, size=(100, 2))
glm = nmo.glm.PopulationGLM(
    regularizer="Ridge", 
    regularizer_strength=dict(
        f1=[[0.1, 0.2], [0.1, 0.2], [0.1, 0.2], [0.1, 0.2], [0.1, 0.2]],
        f2=[[0.1, 0.2], [0.1, 0.2]]
    )
)
glm.fit(X, y)
```

## Backwards compatibility

Most user code is unaffected, if it passed scalar strengths.
Internal APIs:
- `get_proximal_operator` and `penalized_loss` signatures changed (adapters updated).
- Proximal operator functions now expect `(x, strength, scaling)` where strength is a pytree matching parameter structure; if you called them directly, update your calls.

## Performance considerations
- When users pass Python floats, we do not transform to JAX, as python floats are a tiny bit faster.
- Operations are vectorized over pytrees and remain JAX-compatible.
- Overhead is minimal and confined to tree-mapping; no algorithmic slowdowns expected.

## Testing
- Tests added for each regularizer class, checking input types, structure (mis)matches, and contents of `filter_kwargs`.
- Tests added for all proximal operators, verifying they take in and return correctly.

## Documentation
- Docstrings have been updated to match the new possible inputs for `regularizer_strength`.
- A how-to-guide was added about regularizing parameters at all levels, it handles all the examples above.

## Related issues
#197

### CODE REVIEW

File: src/nemos/proximal_operator.py [hunk 1]
```diff
@@ -210,21 +261,10 @@ def prox_elastic_net(
...
-        if isinstance(hyperparams[0], float)
-        else hyperparams[0]
-    )
-    # hyperparams[1] = (1 - regularizer_ratio)/regularizer_ratio
-    gam = (
-        tree_util.tree_map(lambda y: hyperparams[1] * jnp.ones_like(y), x)
-        if isinstance(hyperparams[1], float)
-        else hyperparams[1]
+    lam = jax.tree_util.tree_map(
+        lambda strength, ratio: strength * ratio, strength, ratio
```
Comment (BalzaniEdoardo) [inline src/nemos/proximal_operator.py]:
```suggestion
        lambda s, r: s * r, strength, ratio
```

I find it less confusing

File: src/nemos/proximal_operator.py [hunk 2]
```diff
@@ -210,21 +261,10 @@ def prox_elastic_net(
...
     )
+    gam = jax.tree_util.tree_map(lambda ratio: (1 - ratio) / ratio, ratio)
```
Comment (BalzaniEdoardo) [inline src/nemos/proximal_operator.py]: same as above :)

File: src/nemos/regularizer.py [hunk 3]
```diff
@@ -144,19 +151,30 @@ def allowed_solvers(self) -> Tuple[str]:
...
 
         Returns
         -------
         :
-            The proximal operator, which typically applies a form of regularization.
+            The proximal operator, applying regularization to the provided parameters.
         """
-        pass
+
+        def prox_op(params, strength, scaling=1.0):
```
Comment (BalzaniEdoardo) [inline src/nemos/regularizer.py]: This will conflict with my changes but should be an easy fix

File: src/nemos/regularizer.py [hunk 4]
```diff
@@ -184,12 +202,12 @@ def _check_loss_output_tuple(output: tuple):
                 "or a tuple with two values, the loss and an auxiliary variable."
             )
 
-    def penalized_loss(self, loss: Callable, regularizer_strength: float) -> Callable:
-        """Return a function for calculating the penalized loss using Lasso regularization."""
+    def penalized_loss(self, loss: Callable, strength: float) -> Callable:
```
Comment (BalzaniEdoardo) [inline src/nemos/regularizer.py]: this will also conflict

File: src/nemos/proximal_operator.py [hunk 5]
```diff
@@ -264,26 +277,16 @@ def prox_elastic_net(
...
-        if isinstance(hyperparams[1], float)
-        else hyperparams[1]
-    )
 
     def prox_l1(u, lambd):
         return jnp.sign(u) * jax.nn.relu(jnp.abs(u) - lambd)
 
-    def fun(u, lambd, gamma):
-        return prox_l1(u, scaling * lambd) / (1.0 + scaling * lambd * gamma)
+    def fun(u, strength):
```
Comment (BalzaniEdoardo) [inline src/nemos/proximal_operator.py]: can you rename `strength` to something else in the internal func, it is confusing since you have the same variable name in the outer scope

File: src/nemos/proximal_operator.py [hunk 6]
```diff
@@ -183,23 +173,53 @@ def prox_group_lasso(
...
+    l2_norm = masked_norm_2(x, mask)  # (n_groups,)
+
+    def prox_leaf(xi, mi, s, l2):
+        factor = jax.nn.relu(1.0 - scaling * s / l2)  # (n_groups,)
+        regularized = mi.sum(axis=0) > 0  # (n_features,)
+        return jnp.where(
+            regularized,
+            xi * jnp.einsum("i, i...->...", factor, mi),
+            xi,
+        )
```
Comment (BalzaniEdoardo) [inline src/nemos/proximal_operator.py:185]: Neat, I knew my PR on the GroupLasso would have made this much easier. Very clean stuff

File: src/nemos/regularizer.py [hunk 7]
```diff
@@ -178,14 +174,9 @@ class Regularizer(Base, abc.ABC):
...
-    _default_solver: str = None
-
-    def __init__(
-        self,
-        **kwargs,
-    ):
-        super().__init__(**kwargs)
+    _allowed_solvers: Tuple[str]
+    _default_solver: str
+    _proximal_operator: Callable
```
Comment (BalzaniEdoardo) [inline src/nemos/regularizer.py:180]: Nice, I think that this class attribute is the right call

File: src/nemos/regularizer.py [hunk 8]
```diff
@@ -262,50 +260,159 @@ def _penalized_loss(params, *args, **kwargs):
...
+            or (isinstance(strength, (np.ndarray, jnp.ndarray)) and strength.ndim == 0)
+        ):
+            scalar = 1.0 if strength is None else float(strength)
+            for where in wheres:
+                subtree = where(params)
+                structured_strength = eqx.tree_at(
+                    where,
+                    structured_strength,
+                    jax.tree_util.tree_map(
+                        lambda p: jnp.full(p.shape, scalar, dtype=float), subtree
```
Comment (BalzaniEdoardo) [inline src/nemos/regularizer.py]: If the regularization is a scalar, you should create a tree with scalars in it. 
The tree-map will work anyways because scalars can be added and multiplied with arrays no problem. Keeping strength scalar when possible will make the computation faster. 

We can also check if having a tree of python floats vs `jax.numpy` scalar arrays makes a difference. I believe that scalars will be treated as concrete values at compile time, so would be hardcoded in compilation. This may even make the code faster. Scalar JAX arrays will be treated as non-static. I would try `timeit` over `glm.update(...)` when regularizer strenght is a scalar vs scalar array.

File: src/nemos/regularizer.py [hunk 9]
```diff
@@ -262,50 +260,159 @@ def _penalized_loss(params, *args, **kwargs):
...
+        for s, where in zip(strengths, wheres):
+            subtree = where(params)
+            param_leaves, treedef = jax.tree_util.tree_flatten(subtree)
+
+            strength_leaves = (
+                [s] * len(param_leaves)
+                if isinstance(s, (np.ndarray, jnp.ndarray)) and s.ndim == 0
+                else jax.tree_util.tree_leaves(s)
+            )
+            if len(strength_leaves) != len(param_leaves):
```
Comment (BalzaniEdoardo) [inline src/nemos/regularizer.py]:
```suggestion
            if len(strength_leaves) != treedef.num_leaves:
```

File: src/nemos/regularizer.py [hunk 10]
```diff
@@ -262,50 +260,159 @@ def _penalized_loss(params, *args, **kwargs):
...
-        return {}
+            validated = [
+                jnp.broadcast_to(jnp.asarray(sl, dtype=float), p.shape)
+                for p, sl in zip(param_leaves, strength_leaves)
+            ]
+            structured_strength = eqx.tree_at(
+                where,
+                structured_strength,
+                jax.tree_util.tree_unflatten(treedef, validated),
+            )
```
Comment (BalzaniEdoardo) [inline src/nemos/regularizer.py]: Similar idea, if strength_leaves are scalar keep them as such.

I would create a utility function that does this
```python
  def _structured_strength(strength_leaf, param_leaf):                                                                                                                
      # Keep Python scalars as-is (may be faster than JAX scalars in some cases)                                                                                      
      if isinstance(strength_leaf, Number):                                                                                                                           
          return strength_leaf                                                                                                                                        
      # Scalar array -> keep as 0-d JAX array                                                                                                                         
      if is_array_like(strength_leaf) and jnp.asarray(strength_leaf).ndim == 0:                                                                                       
          return jnp.asarray(strength_leaf, dtype=float)                                                                                                              
      # Non-scalar -> broadcast to param shape                                                                                                                        
      return jnp.broadcast_to(jnp.asarray(strength_leaf, dtype=float), param_leaf.shape)                                                                              
                                                                                         
```

File: src/nemos/regularizer.py [hunk 11]
```diff
@@ -442,62 +508,41 @@ class Lasso(Regularizer):
...
+    def _penalty_on_subtree(self, subtree, strength: Any, **kwargs) -> jnp.ndarray:
         """
         Compute the Lasso penalization for given parameters.
 
         Parameters
         ----------
-        sub_params :
+        subtree :
             Model parameters for which to compute the penalization.
+        substrength :
```
Comment (BalzaniEdoardo) [inline src/nemos/regularizer.py]: this doesn't match the signature

File: src/nemos/regularizer.py [hunk 12]
```diff
@@ -442,62 +508,41 @@ class Lasso(Regularizer):
...
 
         Returns
         -------
         float
             The Lasso penalization value.
         """
 
-        def l1_penalty(coeff: jnp.ndarray) -> jnp.ndarray:
-            return strength * jnp.sum(jnp.abs(coeff))
+        def l1_penalty(coeff: jnp.ndarray, strength: jnp.ndarray):
```
Comment (BalzaniEdoardo) [inline src/nemos/regularizer.py]: same thing with the var naming, can you change this var name here, I am always confused when vars have the same names in different scopes.

File: src/nemos/regularizer.py (hunk 8 above)
Comment (wulfdewolf) [reply src/nemos/regularizer.py]: I checked and indeed using python scalars is a tiiiny bit faster.

**Review by BalzaniEdoardo — APPROVED** [review]
LGTM, you can merge this one! great job, this is going to be very useful

File: docs/how_to_guide/finegrained_regularization.md [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+By default, [`GroupLasso`](nemos.regularizer.GroupLasso) generates masks that match the group stucture in the design matrix `X`.
+If you pass your own mask, you need to make sure the regularizer strength matches its structure.
+
+### PopulationGLM
+A [`PopulationGLM`](nemos.glm.PopulationGLM) models many neurons simultaneously. 
+Internally, this means it will have a set of parameters per neuron.
+All regularization strategies above work for a [`PopulationGLM`](nemos.glm.PopulationGLM) as well:
+```{code-cell} ipython3
+# we'll create a second neuron with twice the amount of spikes
+spikes = jnp.stack([spikes, rate*2], axis=1)
```
Comment (BalzaniEdoardo) [inline docs/how_to_guide/finegrained_regularization.md]: should you generate the counts here instead of the rate?

File: docs/how_to_guide/finegrained_regularization.md [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+glm = nmo.glm.PopulationGLM(
+    regularizer="Ridge", 
+    regularizer_strength=dict(
+        f1=[[0.1, 0.2], [0.1, 0.2], [0.1, 0.2], [0.1, 0.2], [0.1, 0.2]],
```
Comment (BalzaniEdoardo) [inline docs/how_to_guide/finegrained_regularization.md:188]: I would add a bit of explanation on why this reg strength has this structure
